server/datastore/mysql/software_title_icons.go

```diff
@@ -0,0 +1,182 @@
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/server/contexts/ctxerr"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/jmoiron/sqlx"
+)
+
+func (ds *Datastore) CreateOrUpdateSoftwareTitleIcon(ctx context.Context, payload *fleet.UploadSoftwareTitleIconPayload) (*fleet.SoftwareTitleIcon, error) {
+	var query string
+	var args []interface{}
+	query = `
+		INSERT INTO software_title_icons (team_id, software_title_id, storage_id, filename)
+		VALUES (?, ?, ?, ?) ON DUPLICATE KEY UPDATE
+		storage_id = VALUES(storage_id), filename = VALUES(filename)
+	`
+	args = []interface{}{payload.TeamID, payload.TitleID, payload.StorageID, payload.Filename}
+
+	_, err := ds.writer(ctx).ExecContext(ctx, query, args...)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "upserting software title icon")
+	}
+
+	return &fleet.SoftwareTitleIcon{
+		TeamID:          payload.TeamID,
+		SoftwareTitleID: payload.TitleID,
+		StorageID:       payload.StorageID,
+		Filename:        payload.Filename,
+	}, nil
+}
+
+func (ds *Datastore) GetSoftwareTitleIcon(ctx context.Context, teamID uint, titleID uint) (*fleet.SoftwareTitleIcon, error) {
+	args := []interface{}{teamID, titleID}
+	query := `
+		SELECT team_id, software_title_id, storage_id, filename
+		FROM software_title_icons
+		WHERE team_id = ? AND software_title_id = ?
+	`
+	var icon fleet.SoftwareTitleIcon
+	err := sqlx.GetContext(ctx, ds.reader(ctx), &icon, query, args...)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, ctxerr.Wrap(ctx, notFound("SoftwareTitleIcon"), "get software title icon")
+		}
+		return nil, ctxerr.Wrap(ctx, err, "get software title icon")
+	}
+
+	return &icon, nil
+}
+
+func (ds *Datastore) HasAccessToExistingIconFile(ctx context.Context, teamID uint, storageID string) (bool, error) {
+	var exists bool
+	query := `
+		SELECT EXISTS(
+			SELECT 1 FROM software_title_icons WHERE storage_id = ? AND team_id = ?
```

Comment:
@iansltx think it's worth adding an index for `team_id, storage_id` on `software_title_icons`? We already have one on `team_id, software_title_id`

Reply:
I don't. This query will use the team ID part of the team ID/SW title ID index anyway, which should get the result set down to hundreds of entries, at which point scanning them all won't be too bad.

Reply:
Also, this isn't quite right.

We need to account for users that:

1. Are global-permissioned
2. Have permissions on multiple teams

Global-search for

```
// check if we already have the installer based on the SHA256 and URL
```

to see how this logic was implemented elsewhere.

Reply:
Ah you are saying this bit https://github.com/fleetdm/fleet/blob/main/ee/server/service/software_installers.go#L1911-L1955